pkg/hsskey/filereader.go

```diff
@@ -0,0 +1,153 @@
+// If the byte represents a printable ASCII character, add it as is to the output string
+// If it is non-printable or non-ASCII, add it as octal notation.
+func toOctalEscapeSequence(b []byte) string {
+	var shortHandNotions = map[string]string{
+		"\011": "\\t",
+		"\015": "\\r",
+	}
+	octalEscape := ""
+	for _, v := range b {
+		octalChar := fmt.Sprintf("\\%03o", v)
+		if v >= 32 && v <= 126 {
+			octalEscape += string(v)
+		} else if notion, ok := shortHandNotions[string(v)]; ok {
+			// DebugString prints these characters as shorthand notations
+			octalEscape += notion
+		} else {
+			octalEscape += octalChar
+		}
+	}
+	return octalEscape
+}
+
+// validateChecksum check if HSS data is valid based on the checksum value stored in the last 4
+// bytes.
+func validateChecksum(data []byte) bool {
+	size := len(data)
+	if size != hostSecretSeedStructSize {
+		return false
+	}
+	// Split the data into the main part and the checksum. Last 4 bytes is used for storing
+	// checksum value.
+	mainPart := data[:size-4]
+	checksumBytes := data[size-4:]
+
+	// Convert the checksum bytes from big-endian.
+	checksum := binary.BigEndian.Uint32(checksumBytes)
+
+	// Compute the checksum of the main part.
+	table := crc32.MakeTable(crc32.Castagnoli)
+	computedChecksum := crc32.Checksum(mainPart, table)
+
+	if computedChecksum != checksum {
+		return false
+	}
+	return true
+}
+
+// deduplicate removes duplicate entries from a slice of byte slices.
+func deduplicate(data [][]byte) [][]byte {
+	seen := make(map[string]bool)
+	result := [][]byte{}
+
+	for _, entry := range data {
+		strEntry := string(entry)
+		if !seen[strEntry] {
+			seen[strEntry] = true
+			result = append(result, entry)
+		}
+	}
+	return result
+}
+
+// getHssEepromPaths takes a glob pattern for EEPROM sysfs paths and return all the matching paths.
+// This function will return error if the provided glob pattern doesn't match any existing path.
+//
+// For example, if hssEepromPattern is "0-007[01]", this function will look for all the paths
+// matching "/sys/bus/i2c/devices/0-007[01]/eeprom"
+func getHssEepromPaths(basePattern string, hssEepromPattern string) ([]string, error) {
+	fullPathGlob := fmt.Sprintf(basePattern, hssEepromPattern)
+	matches, err := filepath.Glob(fullPathGlob)
+	if err != nil {
+		return nil, fmt.Errorf("failed to search the file path %v", err)
+	}
+	if len(matches) == 0 {
+		return nil, fmt.Errorf("no matching path found for glob pattern %s", fullPathGlob)
+	}
+	return matches, nil
+}
+
+// readFromFile reads HSS keys from the specified file. Expecting the file containing 4 consecutive
+// HSS keys. Each HSS key is 64 bytes long and has a checksum for validation.
+func readHssFromFile(filePath string) ([][]byte, error) {
+	data, err := os.ReadFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file %v", err)
+	}
+	if len(data) < hostSecretSeedStructSize*hostSecretSeedCount {
```

Comment:
Should this be `if len(data) != hostSecretSeedStructSize*hostSecretSeedCount ` ? as we expect exact amount of bytes. 

Or there is a possibility we might read more bytes ? and that is ok(?)

Reply:
we only care about [0, hostSecretSeedStructSize*hostSecretSeedCount] bytes in the storage device. In the ideal case the length of data should be exactly equal to hostSecretSeedStructSize*hostSecretSeedCount, but in case there is more data than expected, we can ignore the rest.

Added a log statement when there are more data than expected.